### PR TITLE
[sdk-metrics] Add experimental envvar for setting ExemplarFilter for histograms

### PR DESCRIPTION
**Pull request by CodeBlanch**
Relates to #2527

## Changes

* Adds an experimental envvar `OTEL_DOTNET_EXPERIMENTAL_METRICS_EXEMPLAR_FILTER_HISTOGRAMS` which can be used to set the default SDK `ExemplarFilter` for histogram metrics.

## Merge requirement checklist

* [X] [CONTRIBUTING](https://github.com/open-telemetry/opentelemetry-dotnet/blob/main/CONTRIBUTING.md) guidelines followed (license requirements, nullable enabled, static analysis, etc.)
* [X] Unit tests added/updated
* [X] Appropriate `CHANGELOG.md` files updated for non-trivial changes

### CODE REVIEW

File: docs/metrics/customizing-the-sdk/README.md [hunk 1]
```diff
@@ -382,6 +385,24 @@ using var meterProvider = Sdk.CreateMeterProviderBuilder()
...
+  using environment variables or configuration.
+
+| Environment variable       | Description                                        | Notes |
+| -------------------------- | -------------------------------------------------- |-------|
+| `OTEL_METRICS_EXEMPLAR_FILTER` | Sets the default `ExemplarFilter` to use for all metrics. | Added in `1.9.0` |
+| `OTEL_DOTNET_EXPERIMENTAL_METRICS_EXEMPLAR_FILTER_HISTOGRAMS`        | Sets the default `ExemplarFilter` to use for histogram metrics. If not set than `OTEL_METRICS_EXEMPLAR_FILTER` also applies to histograms. | Experimental key (may be removed or changed in the future). Added in `1.9.0` |
+
+Allowed values:
+
+* `always_off`: Eqivalent to `ExemplarFilterType.AlwaysOff`
```
Comment (vishweshbankwar) [inline docs/metrics/customizing-the-sdk/README.md:402]:
```suggestion
* `always_off`: Equivalent to `ExemplarFilterType.AlwaysOff`
```

File: docs/metrics/customizing-the-sdk/README.md [hunk 2]
```diff
@@ -382,6 +385,24 @@ using var meterProvider = Sdk.CreateMeterProviderBuilder()
...
+environmental variables:
+
+> [!NOTE]
+> Programmatically calling `SetExemplarFilter` will override any defaults set
+  using environment variables or configuration.
+
+| Environment variable       | Description                                        | Notes |
+| -------------------------- | -------------------------------------------------- |-------|
+| `OTEL_METRICS_EXEMPLAR_FILTER` | Sets the default `ExemplarFilter` to use for all metrics. | Added in `1.9.0` |
+| `OTEL_DOTNET_EXPERIMENTAL_METRICS_EXEMPLAR_FILTER_HISTOGRAMS`        | Sets the default `ExemplarFilter` to use for histogram metrics. If not set than `OTEL_METRICS_EXEMPLAR_FILTER` also applies to histograms. | Experimental key (may be removed or changed in the future). Added in `1.9.0` |
```
Comment (vishweshbankwar) [inline docs/metrics/customizing-the-sdk/README.md:398]: > If not set than `OTEL_METRICS_EXEMPLAR_FILTER` also applies to histograms

This is slightly confusing. Could we say something like `this takes precedence over OTEL_METRICS_EXEMPLAR_FILTER`

File: test/OpenTelemetry.Tests/Metrics/MetricExemplarTests.cs [hunk 3]
```diff
@@ -52,6 +53,14 @@ public class MetricExemplarTests : MetricTestsBase
 
         Assert.NotNull(meterProviderSdk);
         Assert.Equal((ExemplarFilterType?)expectedValue, meterProviderSdk.ExemplarFilter);
+        if (programmaticValue.HasValue)
```
Comment (vishweshbankwar) [inline test/OpenTelemetry.Tests/Metrics/MetricExemplarTests.cs:56]: Could we add a test to validate the precedence between `OTEL_METRICS_EXEMPLAR_FILTER` and `OTEL_DOTNET_EXPERIMENTAL_METRICS_EXEMPLAR_FILTER_HISTOGRAMS`

File: docs/metrics/customizing-the-sdk/README.md [hunk 4]
```diff
@@ -372,6 +372,9 @@ OpenTelemetry SDK comes with the following `ExemplarFilters` (defined on
 The `SetExemplarFilter` extension method on `MeterProviderBuilder` can be used
 to set the desired `ExemplarFilterType` and enable `Exemplar` collection:
 
+> [!NOTE]
+> The `SetExemplarFilter` API was added in the `1.9.0` release.
```
Comment (cijothomas) [inline docs/metrics/customizing-the-sdk/README.md:376]: 1.9.0 is not released, right?

File: docs/metrics/customizing-the-sdk/README.md (hunk 2 above)
Comment (cijothomas) [inline docs/metrics/customizing-the-sdk/README.md:398]:
```suggestion
| `OTEL_DOTNET_EXPERIMENTAL_METRICS_EXEMPLAR_FILTER_HISTOGRAMS`        | Sets the default `ExemplarFilter` to use for histogram metrics. If not set, then `OTEL_METRICS_EXEMPLAR_FILTER` also applies to histograms. | Experimental key (may be removed or changed in the future). Added in `1.9.0` |
```

File: docs/metrics/customizing-the-sdk/README.md (hunk 4 above)
Comment (cijothomas) [inline docs/metrics/customizing-the-sdk/README.md:376]: can we document the default as well? i.e make it explicit that `AlwaysOff` is the default.

Comment (cijothomas) [reply docs/metrics/customizing-the-sdk/README.md:376]: something like:
The default ExemplarFilter is AlwaysOff i.e Exemplars feature is off by default.
You can enable it by using SetExemplarFilter, which will enable Exemplars for **all type** of metrics.

There is an experimental ENV variable, which allow you to enable Exemplars for Histogram types only, given exemplars are the most valuable for histograms.

OTEL_DOTNET_EXPERIMENTAL_METRICS_EXEMPLAR_FILTER_HISTOGRAMS=trace-based will enable Exemplars for Histograms only.

Note that there is some cpu/memory/network overhead to compute/store/export Exemplars, which is why, Exemplars are disabled by default.

Comment (cijothomas) [reply docs/metrics/customizing-the-sdk/README.md:376]: And mention that - there is ability to turn on/off Exemplar for each individual metric today. (if that is allowed via experimental feature, we can mention that here.)

File: docs/metrics/customizing-the-sdk/README.md (hunk 2 above)
Comment (CodeBlanch) [reply docs/metrics/customizing-the-sdk/README.md]: Updated

File: docs/metrics/customizing-the-sdk/README.md (hunk 4 above)
Comment (CodeBlanch) [reply docs/metrics/customizing-the-sdk/README.md:387]: Correct. I figured we would put it out soon (alpha/beta) after this lands so it wouldn't be a big deal to mention it here a bit early. What you think?

File: docs/metrics/customizing-the-sdk/README.md (hunk 2 above)
Comment (CodeBlanch) [reply docs/metrics/customizing-the-sdk/README.md]: I changed this a bit for @vishweshbankwar's comment above.

File: docs/metrics/customizing-the-sdk/README.md (hunk 4 above)
Comment (CodeBlanch) [reply docs/metrics/customizing-the-sdk/README.md:387]: I pushed some changes: https://github.com/open-telemetry/opentelemetry-dotnet/pull/5611/commits/5aafc2f2c006a35d759d2942a2bc35db396f1dcb

LMK

> And mention that - there is ability to turn on/off Exemplar for each individual metric today. (if that is allowed via experimental feature, we can mention that here.)

I'm going to do a different PR which adds docs for the experimental `ExemplarReservoir` stuff and shows how to use the view api to set a custom reservoir.

File: test/OpenTelemetry.Tests/Metrics/MetricExemplarTests.cs (hunk 3 above)
Comment (CodeBlanch) [reply test/OpenTelemetry.Tests/Metrics/MetricExemplarTests.cs:56]: I tweaked this test to verify the histogram key takes precedence over the global one: https://github.com/open-telemetry/opentelemetry-dotnet/pull/5611/commits/50f3eddff516d41397bb59dc3bf7b153dacc93f2

File: docs/metrics/customizing-the-sdk/README.md [hunk 5]
```diff
@@ -352,26 +352,40 @@ tutorial](../exemplars/README.md) demonstrates how to use exemplars to achieve
...
 
-OpenTelemetry SDK comes with the following `ExemplarFilters` (defined on
+OpenTelemetry SDK comes with the following `ExemplarFilter`s (defined on
 `ExemplarFilterType`):
 
-* `AlwaysOff`: Makes no measurements eligible for becoming an `Exemplar`. Using
-  this is as good as turning off the `Exemplar` feature and is the current
-  default.
+* (Default behavior) `AlwaysOff`: Makes no measurements eligible for becoming an
+  `Exemplar`. Using this disables `Exemplar` collection and avoids all
```
Comment (cijothomas) [inline docs/metrics/customizing-the-sdk/README.md:377]:
```suggestion
  `Exemplar`. Using this disables `Exemplar` collection and avoids all exemplar related
```

File: src/OpenTelemetry/Metrics/MeterProviderSdk.cs [hunk 6]
```diff
@@ -490,37 +493,70 @@ private void ApplySpecificationConfigurationKeys(IConfiguration configuration)
...
             {
                 OpenTelemetrySdkEventSource.Log.MeterProviderSdkEvent(
                     $"Exemplar filter configuration value '{configValue}' has been ignored because a value '{this.ExemplarFilter}' was set programmatically.");
                 return;
             }
 
-            ExemplarFilterType? exemplarFilter;
+            if (!TryParseExemplarFilterFromConfigurationValue(configValue, out var exemplarFilter))
+            {
+                OpenTelemetrySdkEventSource.Log.MeterProviderSdkEvent($"Exemplar filter configuration was found but the value '{configValue}' is invalid and will be ignored.");
```
Comment (cijothomas) [inline src/OpenTelemetry/Metrics/MeterProviderSdk.cs:509]: should we be showing string interpolation in our own logging.. though this is just startup only!

File: src/OpenTelemetry/Metrics/MeterProviderSdk.cs [hunk 7]
```diff
@@ -490,37 +493,70 @@ private void ApplySpecificationConfigurationKeys(IConfiguration configuration)
...
+
+            if (!TryParseExemplarFilterFromConfigurationValue(configValue, out var exemplarFilter))
+            {
+                OpenTelemetrySdkEventSource.Log.MeterProviderSdkEvent($"Exemplar filter histogram configuration was found but the value '{configValue}' is invalid and will be ignored.");
+                return;
+            }
+
+            this.ExemplarFilterForHistograms = exemplarFilter;
+
+            OpenTelemetrySdkEventSource.Log.MeterProviderSdkEvent($"Exemplar filter for histograms set to '{exemplarFilter}' from configuration.");
```
Comment (cijothomas) [inline src/OpenTelemetry/Metrics/MeterProviderSdk.cs:535]: i think we should be define dedicated events announcing/logging the actual ExemplarFilter enabled.

**Review by cijothomas — APPROVED** [review]
left couple nits about using dedicated internal logs about which ExemplarFilter is finally enabled.

File: src/OpenTelemetry/Metrics/MeterProviderSdk.cs (hunk 7 above)
Comment (CodeBlanch) [reply src/OpenTelemetry/Metrics/MeterProviderSdk.cs:538]: I added this: https://github.com/open-telemetry/opentelemetry-dotnet/pull/5611/commits/a088d08efc09db9ec00d7a4e41f3b67aee0baa8a

**Review by rajkumar-rangaraj — APPROVED** [review]

**Pull request merged**